bukuserver/forms.py

```diff
@@ -1,18 +1,52 @@
 """Forms module."""
 # pylint: disable=too-few-public-methods, missing-docstring
-from typing import Any, Dict, Tuple
+import re
 from flask_wtf import FlaskForm
-from wtforms.fields import BooleanField, FieldList, StringField, TextAreaField, HiddenField
-from wtforms.validators import DataRequired, InputRequired, ValidationError
-from buku import DELIM, parse_tags
+from wtforms import Form
+from wtforms.fields import BooleanField, FieldList, StringField, TextAreaField, HiddenField, SelectMultipleField
+from wtforms.validators import DataRequired, InputRequired, Length, Regexp, StopValidation
+from buku import DELIM, taglist_str
 from bukuserver import _, _l, LazyString
-from bukuserver.response import Response
 
-def validate_tag(form, field):
+_parse_bool = lambda x: str(x).lower() == 'true'
+
+TAG_RE = re.compile(r'^[^,]*[^,\s]+[^,]*$')
+
+def optional_none(form, field):
+    if field.data is None:
+        raise StopValidation()
+
+def is_string(form, field):
     if not isinstance(field.data, str):
-        raise ValidationError(_('Tag must be a string.'))
-    if DELIM in field.data:
-        raise ValidationError(_('Tag must not contain delimiter "%(delim)s".', delim=DELIM))
+        raise StopValidation(_('The value must be a string.'))
+
+validate_tag = [is_string, Regexp(TAG_RE)]
+
+
+class ValueList(SelectMultipleField):
+    """A form field model for simple value lists, capable of processing regular array data."""
+
+    def __init__(self, *args, item_validators=[], **kwargs):
+        self.data, self._valid, self._field = None, True, StringField(validators=item_validators)
+        super().__init__(*args, choices=[], validate_choice=False, coerce=(lambda x: x), **kwargs)
+
+    def process_data(self, value):
+        self._valid = isinstance(value, (list, tuple, set, type(None)))  # i.e. for JSON input
+        self.data = None
+        if self._valid:
+            super().process_data(value)
+
+    def pre_validate(self, form):
+        _errors = []
+        _field = self._field.bind(form=form, name=self.name, _meta=self.meta, translations=self._translations)  # pylint: disable=no-member
+        for item in (self.data or []):
+            _field.data = item
+            _field.validate(form)
+            _errors += [_field.errors]
+        if any(x for x in _errors):
+            self.errors += _errors
+        if not self._valid:
+            raise StopValidation(self.gettext('Invalid input.'))
```

Comment:
This is a _major_ improvement over how sequential data was handled during JSON parsing originally